### ISSUE TITLE
Bug while typing recursive struct definition with UnionAll in type

### ISSUE THREAD

**Issue by NHDaly** [JuliaLang/julia#35416]
I'm pretty sure this is a bug in the julia subtyping within recursive structs, which I encountered when making a small change to someone else's code:

They currently had their `Node` parameterized to accept any subtype of the _value_ type `V`, but didn't allow subtyped _keys_. So i tried to fix this by changing `K` to `<:K`, but as you'll see below that somehow breaks julia's ability to define the type! 😱 

```julia
julia> module M
       struct Tip end

       struct Node{T,K,X}
         size::Int64
         left::T
         key::K
         value::X
         right::T
       end

       struct AVL{K,V}
         avl:: Union{Tip,Node{AVL{K,V},K,<:V}}  # Works okay with just one subtype
       end

       end # module
WARNING: replacing module M.
Main.M

julia> module M
       struct Tip end

       struct Node{T,K,X}
         size::Int64
         left::T
         key::K
         value::X
         right::T
       end

       struct AVL{K,V}
         avl:: Union{Tip,Node{AVL{K,V},<:K,<:V}}  # Breaks with two subtypes! 😱
       end

       end # module
WARNING: replacing module M.
ERROR: UndefVarError: AVL not defined
Stacktrace:
 [1] top-level scope at REPL[22]:12
 [2] eval(::Module, ::Any) at ./boot.jl:331
 [3] eval_user_input(::Any, ::REPL.REPLBackend) at /Users/daly/builds/julia-1.4/usr/share/julia/stdlib/v1.4/REPL/src/REPL.jl:86
 [4] run_backend(::REPL.REPLBackend) at /Users/daly/.julia/packages/Revise/Pcs5V/src/Revise.jl:1073
 [5] top-level scope at none:0
```

I tried many variants of that definition, including things like `Node{AVL{K,V}, S,T} where {S<:K, T<:V}` and other things, but they all fail with the same error.

Thanks! :)

**Comment by NHDaly** [comment]
I managed to work-around the problem by moving the type assertion to the constructor, which seems like probably good enough for me, but I'm not sure if this will mean that functions specializing on an instance of `AVL{K,V}` won't be able to know for sure that the types of the `Node` field are `<:K` and `<:V`?

```julia
julia> module M
       struct Tip end

       struct Node{T,K,X}
         size::Int64
         left::T
         key::K
         value::X
         right::T
       end

       struct AVL{K,V}
           avl:: Union{Tip,Node{AVL{K,V}}}
           # This constructor is needed to enforce the typing constraints on the Node, which
           # currently julia cannot compute in the field type itself, above.
           # See: https://github.com/JuliaLang/julia/issues/35416
           AVL{K,V}(v::Union{Tip,Node{AVL{K,V},<:K,<:V}}) where {K,V} = new{K,V}(v)
       end

       end # module
WARNING: replacing module M.
Main.M
```

**Comment by JeffBezanson** [comment]
This should eventually be fixed by #33553.
